### ISSUE TITLE
dotnet xunit fails with TestServer when MVC controller returns structured data

### ISSUE THREAD

**Issue by snerks** [xunit/xunit#1499]
Repro project:
https://github.com/snerks/CustomerWebApi

Running a test that uses `Microsoft.AspNetCore.TestHost`, via `dotnet test`, produces a successful outcome.
Running the same test, via `dotnet xunit`, produces a test failure.

The failure under `dotnet xunit` occurs at the following line:
`response.EnsureSuccessStatusCode();`

` System.Net.Http.HttpRequestException : Response status code does not indicate success: 404 (Not Found).`

Project name: `CustomerWebApi.IntegrationTest`
Test name: `Customers_Get_with_empty_db_returns_correct_result`

The Solution also contains a Unit Test Project, with similar dependencies to the Integration Tests, apart from `Microsoft.AspNetCore.TestHost`. I see no difference in results in that Project.

**Environment**
Windows 7 SP1 (64-bit)
Visual Studio Code - Terminal (or `CMD`)

**Comment by bradwilson** [comment]
I've been spending quite a bit of time here, but I am having a hard time tracking down what the actual issue is.

This sample project works fine with `dotnet xunit`: https://github.com/aspnet/Docs/tree/master/aspnetcore/testing/integration-testing/sample

**Comment by bradwilson** [comment]
The problem comes down to the fact that, from a controller, this works:

```csharp
return Ok("Hello, world!");
```

But this does not:

```csharp
return Ok(complexModel);
```

I'm going to have to enlist some of the ASP.NET team to help me understand why that might be, because I don't see anything obvious in any of my diagnostics in xUnit itself.

**Comment by bradwilson** [comment]
@eilon Do you know someone who can help me track this down? I have a simplified repro project here: https://github.com/bradwilson/dotnet-xunit-mvc-repro

**Comment by Eilon** [comment]
This is indeed weird... @javiercn can you take a look? Maybe something with conneg is acting weird in this case?

**Comment by bradwilson** [comment]
Ping @javiercn, any idea what's going on here?

**Comment by javiercn** [comment]
I started looking into this last week but got sidetracked. Something weird is happening as I'm able to hit the endpoint and debug it but somehow the response gets converted to a 404 after the action executes.
In general there are a few things that are usually required for MVC to run E2E test correctly, but usually these things are only related to Views, which is what is tripping me off.

These things are:
* Copy .deps file of the project to the bin folder of the test project.
* Setup the content root to point to the project folder.
* Disable shadows copying so that tests execute from the bin folder.

None of the requirements described above should cause any issue with your tests, so I'm not sure what's going on.

I'm going to investigate this a bit further until I find the root cause. 

On our next planned release we have a library that streamlines all this for you so you don't have to copy the deps file or the content root. (You still need to disable shadow copying).

**Comment by bradwilson** [comment]
> Something weird is happening as I'm able to hit the endpoint and debug it but somehow the response gets converted to a 404 after the action executes.

I concur, I saw the same behavior. It feels like something to do with content negotiation/conversion, though I can't imagine why that would be going sideways. ☹️

**Comment by javiercn** [comment]
I found it. It's actually not something I was aware of, but there is whats happening.
The version of JSON.NET being loaded is the one that xUnit uses, not the one that the app requires. This triggers a missing method exception when the object is being serialized to the response.

This line on startup is capturing that exception and we are seeing a 404 when the exception handler redirects to that route to render the error page.

```
app.UseExceptionHandler("/Home/Error");
```

This is the actual exception being thrown. The version of newtonsoft.json on the app is 9.0.X.

```c#
      System.MissingMethodException : Method not found: 'Void Newtonsoft.Json.JsonWriter.set_AutoCompleteOnClose(Boolean)'.
      Stack Trace:
           at Microsoft.AspNetCore.Mvc.Formatters.JsonOutputFormatter.CreateJsonWriter(TextWriter writer)
           at Microsoft.AspNetCore.Mvc.Formatters.JsonOutputFormatter.WriteObject(TextWriter writer, Object value)
           at Microsoft.AspNetCore.Mvc.Formatters.JsonOutputFormatter.<WriteResponseBodyAsync>d__9.MoveNext()
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
           at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.<InvokeResultAsync>d__19.MoveNext()
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
           at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.<InvokeNextResultFilterAsync>d__24.MoveNext()
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
           at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Rethrow(ResultExecutedContext context)
           at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
           at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.<InvokeNextResourceFilter>d__22.MoveNext()
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
           at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Rethrow(ResourceExecutedContext context)
           at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
           at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.<InvokeFilterPipelineAsync>d__17.MoveNext()
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
           at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.<InvokeAsync>d__15.MoveNext()
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
           at Microsoft.AspNetCore.Builder.RouterMiddleware.<Invoke>d__4.MoveNext()
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
           at Microsoft.AspNetCore.Hosting.Internal.RequestServicesContainerMiddleware.<Invoke>d__3.MoveNext()
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
           at Microsoft.AspNetCore.TestHost.ClientHandler.<>c__DisplayClass3_0.<<SendAsync>b__0>d.MoveNext()
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
           at Microsoft.AspNetCore.TestHost.ClientHandler.<SendAsync>d__3.MoveNext()
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
           at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
           at System.Net.Http.HttpClient.<FinishSendAsyncBuffered>d__58.MoveNext()
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
        C:\work\repros\dotnet-xunit-mvc-repro\WebApp.IntegrationTests\HomeControllerTests.cs(40,0): at HomeControllerTests.<ThisFails>d__1.MoveNext()
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
  Finished:    WebApp.IntegrationTests
```

I can't seem to be able to make xunit use the Json version of the app. I tried making the dependency explicitly on the app and `<AutoGenerateBindingRedirects>true</AutoGenerateBindingRedirects>` without success.

@bradwilson Any idea?

**Comment by bradwilson** [comment]
If that's what it is, then it should already be fixed in 2.3.1 RC 2.

**Comment by bradwilson** [comment]
Yep, I can verify that this is indeed fixed in 2.3.1 RC 2. Thanks for the investigation, @javiercn!

**Comment by snerks** [comment]
Just to confirm - the issue is fixed for me with v2.3.1.

Thanks everyone!
